### PR TITLE
feat: Support iOS 17 text content types

### PR DESCRIPTION
**Pull request by robwalkerco**
## Summary:

While setting up a credit card form in RN, I discovered that iOS 17 supports a number of new content types ([`UITextContentType` docs](https://developer.apple.com/documentation/uikit/uitextcontenttype?language=objc)). In the docs these are marked with the `Beta` flag.

Setting up the new content types is relatively straightforward, but a change is required in https://github.com/facebook/react-native-deprecated-modules to update the `TextInput` prop types. ~~I will open a PR in that repo shortly.~~ I have [opened a PR](https://github.com/facebook/react-native-deprecated-modules/pull/23) to update the prop types. ~~Once that PR is merged, a version bump for that dependency will need to be added to this PR.~~ The PR is merged and the dependency in this PR has been updated.

## Test Plan:

The `rn-tester` app builds and runs successfully. I have added a few new examples of inputs using the new text content types.

### CODE REVIEW

**Comment by robwalkerco** [comment]
Hmm, I think that something in the CI or in the `rn-tester` project config needs to be updated to handle iOS 17

**Comment by robwalkerco** [comment]
> Hmm, I think that something in the CI or in the `rn-tester` project config needs to be updated to handle iOS 17

Looks like the circleci config uses xcode 14.3 https://github.com/facebook/react-native/blob/44a507fdcf9068df40c0debbbf317b0c64988ad4/.circleci/config.yml#L49C14-L49C14

15.0.0 is supported on circleci https://circleci.com/docs/testing-ios/#supported-xcode-versions

**Comment by robwalkerco** [comment]
> Hmm, I think that something in the CI or in the `rn-tester` project config needs to be updated to handle iOS 17

Also, thinking about this, I'm not sure this change could be merged until XCode 15 is at a stable release, it would be required to build these changes, and even then it may be desirable to have a gap between XCode 15 release and the requirement to build RN apps using it 🤔

File: packages/react-native/Libraries/Text/TextInput/RCTBaseTextInputView.m [hunk 1]
```diff
@@ -264,7 +264,27 @@ - (void)setTextContentType:(NSString *)type
       @"password" : UITextContentTypePassword,
       @"newPassword" : UITextContentTypeNewPassword,
       @"oneTimeCode" : UITextContentTypeOneTimeCode,
-    };
+    } mutableCopy];
+
+    if (@available(iOS 17.0, *)) {
```
Comment (philIip) [inline packages/react-native/Libraries/Text/TextInput/RCTBaseTextInputView.m:269]: let's add compile time check here too

__IPHONE_OS_VERSION_MAX_ALLOWED >= __IPHONE_17_0

File: packages/react-native/React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputUtils.mm [hunk 2]
```diff
@@ -211,7 +211,27 @@ UITextContentType RCTUITextContentTypeFromString(std::string const &contentType)
       @"password" : UITextContentTypePassword,
       @"newPassword" : UITextContentTypeNewPassword,
       @"oneTimeCode" : UITextContentTypeOneTimeCode,
-    };
+    } mutableCopy];
+
+    if (@available(iOS 17.0, *)) {
```
Comment (philIip) [inline packages/react-native/React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputUtils.mm:216]: ditto

**Review by philIip — CHANGES_REQUESTED** [review]
i think it's ok to prepare this before xcode 15 rc is out, the change is benign enough.

**Comment by robwalkerco** [comment]
@philIip I've added compile time checks. 🤞 the circleci checks pass now.

The last remaining known issue is the version of `react-native-deprecated-modules` that is being used.
Once the recently merged PR there is published in a new version, I'll update the version in react-native.

Would you like me to prep a PR that bumps the minor version number and updates the changelog of `react-native-deprecated-modules`?

**Comment by robwalkerco** [comment]
> Would you like me to prep a PR that bumps the minor version number and updates the changelog of `react-native-deprecated-modules`?

In an effort to be helpful, I've [opened a PR](https://github.com/facebook/react-native-deprecated-modules/pull/24) on `react-native-deprecated-modules` to bump the version and update the changelog. Once merged and released I'll make the final change on the PR to use the new `react-native-deprecated-modules` version.

File: packages/react-native/Libraries/Text/TextInput/RCTBaseTextInputView.m (hunk 1 above)
Comment (robwalkerco) [reply packages/react-native/Libraries/Text/TextInput/RCTBaseTextInputView.m]: Added, which has resolved the test issues

**Comment by robwalkerco** [comment]
PR for the docs website update - https://github.com/facebook/react-native-website/pull/3804

**Pull request closed without merging**